src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -156,39 +157,63 @@ module IncrementalBuildSyntaxTree =
                 fileParsed.Trigger fileName
 
                 input, sourceRange, fileName, diagnosticsLogger.GetDiagnostics()
+            )
 
+        let parse (parseTask: Threading.Tasks.Task<_>) =
+            try
+                parseTask.Result
             with exn ->
                 let msg = sprintf "unexpected failure in SyntaxTree.parse\nerror = %s" (exn.ToString())
                 System.Diagnostics.Debug.Assert(false, msg)
                 failwith msg
 
-        let parseOrSkip sigNameOpt =
+        // We control access to the cache to prevent theoretical possibility of starting the same parse job more than once.
+        // The semaphore instance here locks single particular cache key, not everything. 
+        let semaphore = new SemaphoreSlim(1)
+        let getValue source =
+            try
+                semaphore.Wait()
```

Comment:
I think the semaphore is only useful when you can use `WaitAsync` so the thread is not blocked. Here the regular lock probably does the same job.

Reply:
Yeah, I thought so too. This function runs fast, it doesn't really wait on anything, just fire-and-forgets the parse task, so it's not worth it to make it async. If the lock blocks, it blocks not for the entire parse duration, but just so it doesn't start the same job twice needlessly.

Reply:
Yeah that should be ok.